sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala

```diff
@@ -730,6 +713,67 @@ class AdaptiveQueryExecSuite
     }
   }
 
+  private def checkSkewJoin(
+      joins: Seq[SortMergeJoinExec],
+      leftSkewNum: Int,
+      rightSkewNum: Int): Unit = {
+    assert(joins.size == 1 && joins.head.isSkewJoin)
+    assert(joins.head.left.collect {
+      case r: CustomShuffleReaderExec => r
+    }.head.partitionSpecs.collect {
+      case p: PartialReducerPartitionSpec => p.reducerIndex
+    }.distinct.length == leftSkewNum)
+    assert(joins.head.right.collect {
+      case r: CustomShuffleReaderExec => r
+    }.head.partitionSpecs.collect {
+      case p: PartialReducerPartitionSpec => p.reducerIndex
+    }.distinct.length == rightSkewNum)
+  }
+
+  test("SPARK-32201: handle general skew join pattern") {
```

Comment:
It's hard to build a test case like the case in PR description. So I just want to give a simple one. But the rule is same. In this test case, we build a SMJ like:
```
SMJ
     Sort
       Shuffle
     Sort
       HashAggregate
         Shuffle
```
It has already breaked current matching pattern. But this test case will fail since below condition:
https://github.com/apache/spark/blob/5d296ed39e3dd79ddb10c68657e773adba40a5e0/sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala#L274-L281
In this test case the HashAggregate will introduce an additional exchange. But this is not related to this PR. They are totally irrelevant.

The plan before patch is
```
AdaptiveSparkPlan(isFinalPlan=true)
+- *(5) SortMergeJoin [key1#183L], [key2#189L], Inner, false
   :- *(3) Sort [key1#183L ASC NULLS FIRST], false, 0
   :  +- CustomShuffleReader coalesced
   :     +- ShuffleQueryStage 0
   :        +- Exchange hashpartitioning(key1#183L, 100), true
   :           +- *(1) Project [CASE WHEN (id#181L < 250) THEN 249 WHEN (id#181L >= 750) THEN 1000 ELSE id#181L END AS key1#183L, id#181L AS value1#184L]
   :              +- *(1) Range (0, 1000, step=1, splits=10)
   +- *(4) Sort [key2#189L ASC NULLS FIRST], false, 0
      +- *(4) HashAggregate(keys=[key2#189L], functions=[sum(value2#190L)], output=[key2#189L, sum2#193L])
         +- CustomShuffleReader coalesced
            +- ShuffleQueryStage 1
               +- Exchange hashpartitioning(key2#189L, 100), true
                  +- *(2) HashAggregate(keys=[key2#189L], functions=[partial_sum(value2#190L)], output=[key2#189L, sum#200L])
                     +- *(2) Project [CASE WHEN (id#187L < 250) THEN 249 ELSE id#187L END AS key2#189L, id#187L AS value2#190L]
                        +- *(2) Range (0, 1000, step=1, splits=10)
```
The plan after is
```
AdaptiveSparkPlan(isFinalPlan=true)
+- *(5) SortMergeJoin(skew=true) [key1#183L], [key2#189L], Inner, true
   :- *(3) Sort [key1#183L ASC NULLS FIRST], false, 0
   :  +- CustomShuffleReader coalesced and skewed
   :     +- ShuffleQueryStage 0
   :        +- Exchange hashpartitioning(key1#183L, 100), true
   :           +- *(1) Project [CASE WHEN (id#181L < 250) THEN 249 WHEN (id#181L >= 750) THEN 1000 ELSE id#181L END AS key1#183L, id#181L AS value1#184L]
   :              +- *(1) Range (0, 1000, step=1, splits=10)
   +- *(4) Sort [key2#189L ASC NULLS FIRST], false, 0
      +- *(4) HashAggregate(keys=[key2#189L], functions=[sum(value2#190L)], output=[key2#189L, sum2#193L])
         +- CustomShuffleReader coalesced
            +- ShuffleQueryStage 1
               +- Exchange hashpartitioning(key2#189L, 100), true
                  +- *(2) HashAggregate(keys=[key2#189L], functions=[partial_sum(value2#190L)], output=[key2#189L, sum#200L])
                     +- *(2) Project [CASE WHEN (id#187L < 250) THEN 249 ELSE id#187L END AS key2#189L, id#187L AS value2#190L]
                        +- *(2) Range (0, 1000, step=1, splits=10)
```

But the ensureRequirements will change to
```
SortMergeJoin(skew=true) [key1#220L], [key2#226L], Inner
:- Sort [key1#220L ASC NULLS FIRST], false, 0
:  +- CustomShuffleReader coalesced and skewed
:     +- ShuffleQueryStage 0
:        +- Exchange hashpartitioning(key1#220L, 100), true, [id=#116]
:           +- *(1) Project [CASE WHEN (id#218L < 250) THEN 249 WHEN (id#218L >= 750) THEN 1000 ELSE id#218L END AS key1#220L, id#218L AS value1#221L]
:              +- *(1) Range (0, 1000, step=1, splits=10)
+- Sort [key2#226L ASC NULLS FIRST], false, 0
   +- HashAggregate(keys=[key2#226L], functions=[sum(value2#227L)], output=[key2#226L, sum2#230L])
      +- Exchange hashpartitioning(key2#226L, 100), true, [id=#172]
         +- CustomShuffleReader coalesced
            +- ShuffleQueryStage 1
               +- Exchange hashpartitioning(key2#226L, 100), true, [id=#127]
                  +- *(2) HashAggregate(keys=[key2#226L], functions=[partial_sum(value2#227L)], output=[key2#226L, sum#237L])
                     +- *(2) Project [CASE WHEN (id#224L < 250) THEN 249 ELSE id#224L END AS key2#226L, id#224L AS value2#227L]
                        +- *(2) Range (0, 1000, step=1, splits=10)
```

Reply:
In our internal Spark, we add a configuration to disable this check to maximize the skew join optimization.
` if (numShuffles > 0 && conf.getConf(SQLConf.SKEW_JOIN_NO_EXTRA_SHUFFLE)) `. So the above UT can work well. But this is another topic. Or we could rewrite a new unit case to avoid this check.